app/src/main/java/org/simple/clinic/patient/PatientRepository.kt

```diff
@@ -67,21 +77,66 @@ class PatientRepository @Inject constructor(
 
     val searchableName = nameToSearchableForm(name)
 
+    val substringSearch = database.patientSearchDao()
+        .search(searchableName, dateOfBirthUpperBound, dateOfBirthLowerBound)
+        .toObservable()
+
     return if (includeFuzzyNameSearch.not()) {
-      database.patientSearchDao()
-          .search(searchableName, dateOfBirthUpperBound, dateOfBirthLowerBound)
-          .toObservable()
+      substringSearch.compose(sortByCurrentFacility())
 
     } else {
       val fuzzySearch = database.fuzzyPatientSearchDao()
           .searchForPatientsWithNameLikeAndAgeWithin(searchableName, dateOfBirthUpperBound, dateOfBirthLowerBound)
           .toObservable()
 
-      database.patientSearchDao()
-          .search(searchableName, dateOfBirthUpperBound, dateOfBirthLowerBound)
-          .toObservable()
+      substringSearch
           .zipWith(fuzzySearch)
           .map { (results, fuzzyResults) -> (fuzzyResults + results).distinctBy { it.uuid } }
+          .compose(sortByCurrentFacility())
+    }
+  }
+
+  // TODO: Get user from caller.
+  private fun sortByCurrentFacility(): ObservableTransformer<List<PatientSearchResult>, List<PatientSearchResult>> {
+    return ObservableTransformer { searchResults ->
+      val currentFacilityUuidStream = userSession
+          .requireLoggedInUser()
+          .switchMap { facilityRepository.currentFacility(it) }
+          .map { it.uuid }
+
+      val patientToFacilityUuidStream = searchResults
+          .map { patients -> patients.map { it.uuid } }
+          .switchMap {
+            database
+                .bloodPressureDao()
+                .patientToFacilityIds(it)
+                .toObservable()
+          }
+
+      Observables.combineLatest(searchResults, currentFacilityUuidStream, patientToFacilityUuidStream)
+          .map { (patients, currentFacility, patientToFacilities) ->
+            val patientToUniqueFacilities = patientToFacilities
+                .fold(mutableMapOf<PatientUuid, MutableSet<FacilityUuid>>()) { facilityUuids, (patientUuid, facilityUuid) ->
+                  if (patientUuid !in facilityUuids) {
+                    facilityUuids[patientUuid] = mutableSetOf()
+                  }
+
+                  facilityUuids[patientUuid]?.add(facilityUuid)
+                  facilityUuids
+                } as Map<PatientUuid, Set<FacilityUuid>>
+
+            patients.sortedWith(Comparator { first, second ->
```

Comment:
Can you change the Comparator to behave as per the spec and use `sortedByDescending` on the collection instead?